src/components/userGuidedForm.tsx

```diff
@@ -0,0 +1,263 @@
+  initialColor?: `#${string}`;
+}
+type AnyObject = { [key: string]: unknown };
+
+// Check if the value is empty
+const isEmptyValue = (value: unknown) => {
+  if (value === null || value === undefined) return true;
+  if (typeof value === 'string' && value.trim() === '') return true;
+  if (Array.isArray(value) && value.length === 0) return true;
+  if (typeof value === 'object' && Object.keys(value).length === 0) return true;
+  return false;
+};
+// Return true if one or more values are empty
+const hasEmptyValues = (obj: AnyObject): boolean => {
+  return Object.values(obj).some(isEmptyValue);
+};
+
+const predefinedKeywords = [
+  'Bright',
+  'Dark',
+  'Fresh',
+  'Highlighted',
+  'Realistic',
+];
+
+export default function UserGuidedForm({
+  initialColor = '#anyColor',
+}: userGuidedFormProps) {
+  const [purpose, setPurpose] = useState<string>('');
+  const [purposeError, setPurposeError] = useState<boolean>(false);
+
+  const [mood, setMood] = useState<string>('');
+  const [moodError, setMoodError] = useState<boolean>(false);
+
+  const [audience, setAudience] = useState<string>('');
+  const [audienceError, setAudienceError] = useState<boolean>(false);
+
+  const [keyword, setKeywords] = useState<string>('');
+  const [selectedKeywords, setSelectedKeywords] = useState<string[]>([]);
+  const [keywordsError, setKeywordsError] = useState<boolean>(false);
+
+  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const { id, value } = event.target;
+
+    switch (id) {
+      case 'purpose':
+        setPurpose(value);
+        setPurposeError(!countChars(event));
+        break;
+      case 'audience':
+        setAudience(value);
+        setAudienceError(!countChars(event));
+        break;
+      case 'mood':
+        setMood(value);
+        setMoodError(!countOneWord(event));
+        break;
+      case 'keywords':
+        setKeywords(value);
+        break;
+      default:
+        break;
+    }
+  };
+
+  const countOneWord = (event: React.ChangeEvent<HTMLInputElement>) => {
+    return event.target.value.split(' ').length === 1;
+  };
+
+  const countChars = (event: React.ChangeEvent<HTMLInputElement>) => {
+    return event.target.value.length <= 255;
+  };
+
+  const handleSelectChange = (value: string) => {
+    if (!selectedKeywords.includes(value) && selectedKeywords.length < 5) {
+      setSelectedKeywords([...selectedKeywords, value]);
+    }
+  };
+
+  const handleAddCustomKeyword = () => {
+    if (
+      keyword &&
+      !selectedKeywords.includes(keyword) &&
+      selectedKeywords.length < 5
+    ) {
+      setSelectedKeywords([...selectedKeywords, keyword]);
+      setKeywords('');
+    } else if (selectedKeywords.length >= 5) {
+      setKeywordsError(true);
+    } else {
+      setKeywordsError(false);
+    }
+  };
+
+  const handleDeleteKeyword = (keyword: string) => {
+    setSelectedKeywords(selectedKeywords.filter((word) => word !== keyword));
+    setKeywordsError(false);
+  };
```

Comment:
When we have such a complex state and validation, is better to use something like useReducer instead of useState, so that we can keep the business logic separated from the presentation. The main problem is that all those helper functions are inside the form component, which will cause all of them to be recreated every time users type in any of the fields, which can cause serious hit in performance, and also it makes the components too much overwelming and hard to manage.

You can check more here: https://joshua777.hashnode.dev/how-to-use-the-usereducer-hook-for-simple-form-validation-in-react

Another alternative is using validation libraries like Zod, so that you keep your component dry, or maybe libraries like react-hook-form.